Render widget groups in the back end with CSS grid

zoglo opened the pull request:
Reviving #2053 with this for Contao ^5.3 and the new introduced classes (w25, w33, w50, w66, w75).
Could be merged with additional changes and would be BC-compatible due to the `widget-group` wrapper.

Keep in mind this is the reworked CSS-Grid based on @m-vo's solution for the backend widgets as discussed in the contao call for #1024.

___

#### Rewrite

Also grouped most media-queries so it's mobile first (not the split-button tho, doesn't make sense for that one)

___

#### Additional classes

Introcuded `.w16` on that occasion

___

#### Custom properties

Following custom properties were added:

| Custom property | Description |
| :- | :- |
| `--widget-col-amount` | Column amount of the backend-grid<br>_(Defaults to 12)_ |
| `--widget-cols` | Amount of space that this takes per row<br>_(Can be used to create more specific widget classes)_ |
| `--widget-gap-rows` | Row gap for widget-group container |
| `--widget-gap-cols` | Column gap for widget-group container |

zoglo committed:
Rewrite #2053 for Contao 5.3 (5.x)

zoglo committed:
Mobile and tablet styles for grid widgets

- See m-vo at 2053

zoglo committed:
Revert editorConfig

zoglo commented:
Questionable if we'd need other classes than `.w25,.w33,.w50,.w66,.w75`.
I changed this to a Draft PR due to possible optimizations with custom properties and not using "max-width" media queries. 

Tbh, @m-vo's PR was flawless, it just needed some love and some edge-cases.

ausi commented on core-bundle/contao/themes/flexible/styles/basic.css:
```diff
@@ -347,10 +347,31 @@ table.with-zebra tbody tr:nth-child(even) td {
+.widget-group {
+	display: grid;
+	grid-column-gap: 30px;
+	grid-template-columns: repeat(12, 1fr);
```
```suggestion
	grid-template-columns: repeat(12, minmax(0, 1fr));
```

Otherwise items that are too wide would blow up the column width.

ausi commented on core-bundle/contao/themes/flexible/styles/basic.css:
```diff
@@ -347,10 +347,31 @@ table.with-zebra tbody tr:nth-child(even) td {
+	grid-column: auto / span 3;
+}
+.widget-group > .clr {
+	grid-column-start: 1 !important;
```
```suggestion
	grid-column-start: 1;
```
We should not use `!important`. 
Instead we should replace all `grid-column: auto / span …;` with `grid-column-end: span …;`

zoglo replied:
Forgot that one after testing, yours is a good solution tho.

Mind that this may change over the next days but I'll prefer yours.

zoglo replied:
Using grid-column-end would lead to grid-column-start not being auto anymore, hence always being end + every widget being automatically in a new row.

Simply removing the `!important` is enough, however I'll move the `.clr`to the end of the css file as well.

zoglo replied:
@ausi 84ca874f0bc3aa1f2af4b01592ff378fac49fc76 should work tho

fritzmg replied:
> Using grid-column-end would lead to grid-column-start not being auto anymore, hence always being end + every widget being automatically in a new row.

It should work without it. See https://codepen.io/fritzmg/pen/WNWJOOV for example.

// edit: see my suggestions below

zoglo replied:
Added with your suggestions

zoglo committed:
Merge branch 'contao:5.x' into feature/grid-widget-backend

zoglo committed:
Add a size range for grid template columns

Co-authored-by: Martin Auswöger <martin@auswoeger.com>

zoglo committed:
Update grid-layout

- do not use column-gap on handhelds (overflow issue)
- do not use !important for .clr

fritzmg commented on core-bundle/contao/themes/flexible/styles/basic.css:
```diff
@@ -1056,13 +1029,34 @@ a.tl_submit {
+	.widget-group > .w50,
+	.widget-group > .w66,
+	.widget-group > .w75 {
+		grid-column: 1 / 13;
```
```suggestion
		grid-column-end: span 12;
```

fritzmg commented on core-bundle/contao/themes/flexible/styles/basic.css:
```diff
@@ -1056,13 +1029,34 @@ a.tl_submit {
+	.widget-group > .w50,
+	.widget-group > .w66,
+	.widget-group > .w75 {
+		grid-column: auto / span 6;
```
```suggestion
		grid-column-end: span 6;
```

fritzmg commented on core-bundle/contao/themes/flexible/styles/basic.css:
```diff
@@ -347,10 +347,28 @@ table.with-zebra tbody tr:nth-child(even) td {
+	grid-template-columns: repeat(12, minmax(0, 1fr));
+}
+.widget-group > * {
+	grid-column: 1 / 13;
```
```suggestion
	grid-column-end: span 12;
```

fritzmg commented on core-bundle/contao/themes/flexible/styles/basic.css:
```diff
@@ -347,10 +347,28 @@ table.with-zebra tbody tr:nth-child(even) td {
+	margin: 0;
+}
+.widget-group > .w50 {
+	grid-column: auto / span 6;
```
```suggestion
	grid-column-end: span 6;
```

fritzmg commented on core-bundle/contao/themes/flexible/styles/basic.css:
```diff
@@ -347,10 +347,28 @@ table.with-zebra tbody tr:nth-child(even) td {
+	grid-column: auto / span 6;
+}
+.widget-group > .w33 {
+	grid-column: auto / span 4;
```
```suggestion
	grid-column-end: span 4;
```

fritzmg commented on core-bundle/contao/themes/flexible/styles/basic.css:
```diff
@@ -347,10 +347,28 @@ table.with-zebra tbody tr:nth-child(even) td {
+	grid-column: auto / span 4;
+}
+.widget-group > .w25 {
+	grid-column: auto / span 3;
```
```suggestion
	grid-column-end: span 3;
```

fritzmg commented on core-bundle/contao/themes/flexible/styles/basic.css:
```diff
@@ -1056,13 +1029,34 @@ a.tl_submit {
+	.widget-group {
+		grid-column-gap: 0;
+	}
+
```
```suggestion
```

This is not necessary if there are no actual columns (all have span 12), or is it? Also the property is called `column-gap` (in CSS3).

zoglo replied:
This one is necessary due to the column-gap being 30px by default and enlarging the row on mobile devices (12x30px + margin + paddings).

We could minmax this but it may lead to other layouts than they have been before.

fritzmg replied:
But on mobile there are no columns, everything is `span 12` - so the `column-gap` isn't applied anywhere.

zoglo replied:
After your latest changes :)

zoglo replied:
![image](https://github.com/contao/contao/assets/55794780/a50d28c5-a0a2-4792-8e9b-882235556f3e)

That's what happens till 410px (due to the margin (10px * 2) + padding (15px * 2) + gap (12 * 30px) ), thus leaving it will fix the issue.
This CSS is already desktop first so resetting it won't hurt and saves one additional media-query.

fritzmg replied:
I am still not sure what you mean. There is no visible difference - as it makes no difference, since all widgets span the full 12 columns. Resetting the `column-gap` to `0` can be removed.

ausi replied:
I think he means that the grid is too wide then (as you can see in the screenshot) as the gaps are always there even if all widgets span across them.

fritzmg replied:
Ah sorry, I see now. After a certain width the column size is `0` and the `column-gap`s create a minimum width.

zoglo committed:
Use `grid-column-end` for widget cols

Co-authored-by: Fritz Michael Gschwantner <fmg@inspiredminds.at>

zoglo committed:
Rebuild the assets

zoglo committed:
Use custom properties for backend widgets and introduce `w16`

- added `--widget-col-amount`
- added `--widget-cols`

zoglo commented:
I think this PR is finished if the solution mentioned in https://github.com/contao/contao/pull/7087#discussion_r1558289824 is fine.

The last commit [Use custom properties for backend widgets and introduce w16-class](https://github.com/contao/contao/pull/7087/commits/608b93e0e63e79368e93c4eed980c41ab6e7d3b0) introduces custom properties so people can modify their backend grid based on the following custom properties _(This doesn't have to apply to the whole backend and can be controlled by adding a specific class to a parent)_.

| Custom property | Description |
| :- | :- |
| `--widget-col-amount` | Column amount of the backend-grid<br>_(Defaults to 12)_ |
| `--widget-cols` | Amount of space that this takes per row<br>_(Can be used to create more specific widget classes)_ |

fritzmg commented:
> This CSS is already desktop

On that note, may be it should be changed to mobile first? This would simplify the grid CSS even more.

fritzmg commented on core-bundle/contao/themes/flexible/styles/basic.css:
```diff
@@ -1083,3 +1074,31 @@ a.tl_submit {
+}
+*:not(.widget-group) > .m12 {
+	margin: 0 15px;
+}
```
The coding style still needs to be adjusted here :) (space after `:` and break after `}`).

zoglo replied:
Fixed in 87c308d23bf6a1d594a280a06b00593682253bd3

fritzmg commented on core-bundle/contao/themes/flexible/styles/basic.css:
```diff
@@ -347,10 +347,33 @@ table.with-zebra tbody tr:nth-child(even) td {
+.widget-group {
+	--widget-col-amount: 12;
+	display: grid;
+	grid-column-gap: 30px;
```
```suggestion
	column-gap: 30px;
```

zoglo replied:
See [below](https://github.com/contao/contao/pull/7087#discussion_r1560054220)

zoglo commented:
> On that note, may be it should be changed to mobile first? This would simplify the grid CSS even more.

Could do that for the grid. I'd prefer having `.split-button` in the max-media-query though

fritzmg commented on core-bundle/contao/themes/flexible/styles/basic.css:
```diff
@@ -1056,13 +1034,26 @@ a.tl_submit {
+@media (max-width: 767.98px) {
+
+	.widget-group {
+		grid-column-gap: 0;
```
```suggestion
		column-gap: 0;
```

zoglo replied:
Rewrote it to:
```css
.widget-group {
	--widget-col-amount: 12;
	--widget-gap-rows: 0;
	--widget-gap-cols: 0;
	display: grid;
	gap: var(--widget-gap-rows) var(--widget-gap-cols);
	grid-template-columns: repeat(var(--widget-col-amount), minmax(0, 1fr));
}
```

This also allows for modification of the row itself if people need it in their own backend widgets

zoglo replied:
#### Custom properties

Added the following ones for more customization, see initial description for the other ones

| Custom property | Description |
| :- | :- |
| `--widget-gap-rows` | Row gap for widget-group container |
| `--widget-gap-cols` | Column gap for widget-group container |

#### Media queries

I also grouped most media-queries so it's mobile first (not the split-button tho, doesn't make sense for that one)

zoglo committed:
Mobile first + CS

- also introduce `--widget-gap-rows` and `--widget-gap-cols`

zoglo commented:
@fritzmg 

I did the CS and did the following changes:

#### Media queries

Grouped all media-queries so it's mobile first (not the split-button tho, doesn't make sense for that one)

#### Custom properties

Added the following ones for more customization, see initial description for the other ones

| Custom property | Description |
| :- | :- |
| `--widget-gap-rows` | Row gap for widget-group container |
| `--widget-gap-cols` | Column gap for widget-group container |

zoglo committed:
Update widget layout

- consider sub-palettes
- consider `non-width` classes (always 100% width)
- reset cbx styles properly

fritzmg reviewed (dismissed):
Noice 👍 

Approving tentatively (haven't tested everything exhaustively.)

leofeyer committed:
Merge branch '5.x' into feature/grid-widget-backend

leofeyer committed:
Revert the unrelated changes

leofeyer commented on core-bundle/contao/themes/flexible/styles/basic.css:
```diff
@@ -927,28 +946,66 @@ a.tl_submit {
+	}
+
+	.cbx.m12 {
+		min-height: 80px;
```
This adds the minimum height again. Shouldn‘t it be `*:not(.widget-group) > .cbx.m12` instead? (Also applies to the other BC classes.)

zoglo replied:
Updated these in the latest commit

leofeyer commented on core-bundle/contao/themes/flexible/styles/basic.css:
```diff
@@ -681,6 +729,11 @@ select[multiple] {
+	margin-top: 1px;
+	margin-bottom: 1px;
+}
+
```
What does this do?

zoglo replied:
That's a remnant from here and has been rewritten to be mobile first:
https://github.com/contao/contao/blob/276e0400f0b3eb1ab9bcd6a5e7b03b9b64319722/core-bundle/contao/themes/flexible/styles/basic.css#L1081-L1084

It's for vertical positioning of checkboxes (such as the CSS framework section within layout) but could technically be removed completely.

zoglo replied:
I guess you got confused about this since this CSS was below the "BC comment".
I fixed the order in the latest commit.

leofeyer replied:
I also think that we can remove the two `.tl_checkbox_container .tl_checkbox` rules (one defining a `1px` margin and the other one removing it again).

zoglo replied:
> I also think that we can remove the two `.tl_checkbox_container .tl_checkbox` rules (one defining a `1px` margin and the other one removing it again).

Removed in 0ad58fbc113964e55fa52de25d7e6666bdd2d43a

leofeyer commented:
I like the concept a lot. Well done!

zoglo committed:
Consider more classes for backwards compatibility

- Also sort BC css

zoglo committed:
Run the tools

leofeyer commented on core-bundle/contao/themes/flexible/styles/basic.css:
```diff
@@ -927,28 +946,69 @@ a.tl_submit {
+
+	.m12 {
+		padding: 18px 0 16px;
+	}
```
Could we omit this entirely if we made `align-items: center;` the default behavior?

zoglo replied:
> Could we omit this entirely if we made `align-items: center;` the default behavior?

We can't, because:
- it would have that annoying 1px offset, hence needing to remove 1px padding from `.tl_box h3, .tl_tbox h3, .tl_xpl h3`
   ![image](https://github.com/contao/contao/assets/55794780/fcb04d66-90ef-47ba-ae2b-f61fda578920)

- if you don't have a label, it would never be centered:
   ![image](https://github.com/contao/contao/assets/55794780/84e6bee4-8843-499d-8f20-2e0c8785f527)

fritzmg replied:
> if you don't have a label, it would never be centered:

imho we can ignore that

zoglo replied:
So we reduce 1px on the tlbox label, then that should be fine as well with the offset

leofeyer replied:
Or we add a 1px top margin to the checkbox containers?

leofeyer replied:
With https://github.com/zoglo/contao/pull/2 it works without the `.m12` class for me.

zoglo replied:
When `.widget-group` isn't present, it would still needs a fallback:
See: 0e6e8ff1974f966430a7577ae565e770c9f86dff

leofeyer committed:
Merge branch '5.x' into feature/grid-widget-backend

leofeyer committed:
Use `align-items: center` instead of `.m12`

zoglo committed:
Use `align-items: center` instead of `.m12`

zoglo committed:
Consider `m12` for backwards compatibility

zoglo committed:
Remove unnecessary margins for `.tl_checkbox`

Co-authored-by: leofeyer <1192057+leofeyer@users.noreply.github.com>

leofeyer committed:
Rebuild the assets

leofeyer commented:
Thank you @zoglo.

fritzmg commented:
🎉
